lib/livebook/intellisense/identifier_matcher.ex

```diff
@@ -0,0 +1,1146 @@
+
+  @exact_matcher &Kernel.==/2
+  @prefix_matcher &String.starts_with?/2
+
+  @doc """
+  Returns a list of identifiers matching the given `hint`
+  together with relevant information.
+
+  Evaluation binding and environment is used to expand aliases,
+  imports, nested maps, etc.
+
+  `hint` may be a single token or line fragment like `if Enum.m`.
+  """
+  @spec completion_identifiers(String.t(), Code.binding(), Macro.Env.t()) ::
+          list(identifier_item())
+  def completion_identifiers(hint, binding, env) do
+    context = cursor_context(hint)
+    ctx = %{binding: binding, env: env, matcher: @prefix_matcher}
+    context_to_matches(context, ctx)
+  end
+
+  @doc """
+  Extracts information about an identifier found in `column`
+  in `line`.
+
+  The function returns range of columns where the identifier
+  is located and a list of matching identifier items.
+  """
+  @spec locate_identifier(String.t(), pos_integer(), Code.binding(), Macro.Env.t()) ::
+          %{
+            matches: list(identifier_item()),
+            range: nil | %{from: pos_integer(), to: pos_integer()}
+          }
+  def locate_identifier(line, column, binding, env) do
+    case surround_context(line, {1, column}) do
+      %{context: context, begin: {_, from}, end: {_, to}} ->
+        ctx = %{binding: binding, env: env, matcher: @exact_matcher}
+        matches = context_to_matches(context, ctx)
+        %{matches: matches, range: %{from: from, to: to}}
+
+      :none ->
+        %{matches: [], range: nil}
+    end
+  end
+
+  # Takes a context returned from Code.Fragment.cursor_context
+  # or Code.Fragment.surround_context and looks up matching
+  # identifier items
+  defp context_to_matches(context, ctx) do
+    case context do
+      {:alias, alias} ->
+        match_alias(List.to_string(alias), ctx)
+
+      {:unquoted_atom, unquoted_atom} ->
+        match_erlang_module(List.to_string(unquoted_atom), ctx)
+
+      {:dot, path, hint} ->
+        match_dot(path, List.to_string(hint), ctx)
+
+      {:dot_arity, path, hint} ->
+        match_dot(path, List.to_string(hint), %{ctx | matcher: @exact_matcher})
+
+      {:dot_call, _path, _hint} ->
+        match_default(ctx)
+
+      :expr ->
+        match_default(ctx)
+
+      {:local_or_var, local_or_var} ->
+        match_local_or_var(List.to_string(local_or_var), ctx)
+
+      {:local_arity, local} ->
+        match_local(List.to_string(local), %{ctx | matcher: @exact_matcher})
+
+      {:local_call, _local} ->
+        match_default(ctx)
+
+      {:operator, operator} ->
+        match_local_or_var(List.to_string(operator), ctx)
+
+      {:operator_arity, operator} ->
+        match_local(List.to_string(operator), %{ctx | matcher: @exact_matcher})
+
+      {:operator_call, _operator} ->
+        match_default(ctx)
+
+      {:module_attribute, attribute} ->
+        match_module_attribute(List.to_string(attribute), ctx)
+
+      # :none
+      _ ->
+        []
+    end
+  end
+
```

Comment:
From this point forward this is pretty much the same as the current `Livebook.Intellisense.Completion`, just changed some naming from `complete_` to `match_`. At the end I pasted the content of `Code.Fragment` with the small change regarding `:elixir_tokenizer.tokenize` as pointed out [here](https://github.com/elixir-lang/elixir/pull/11143/files#r676519050).